File: bitvm/merkle/write.js
import { trailingZeros } from '../../libs/common.js'
import { bit_state, bit_state_commit, bit_state_unlock } from '../../scripts/opcodes/u32_state.js'
import { u32_toaltstack, u32_fromaltstack } from '../../scripts/opcodes/u32_std.js'
import {
    u160_state_commit,
    u160_state_unlock,
    u160_state,
    u160_equalverify,
    u160_push,
    u160_swap_endian,
    u160_toaltstack,
    u160_fromaltstack,
    u160_notequal,
} from '../../scripts/opcodes/u160_std.js'
import { blake3_160 } from '../../scripts/opcodes/blake3.js'
import { Leaf, TimeoutLeaf, Transaction, EndTransaction } from '../../scripts/transaction.js'
import {
    LOG_TRACE_LEN,
    LOG_PATH_LEN,
    PATH_LEN,
    VICKY,
    PAUL,
    TIMEOUT
} from '../constants.js'


export class MerkleChallengeCLeaf extends Leaf {

    lock(vicky, paul, merkleIndex) {
        return [
            merkleIndex, OP_DROP, // This is just a marker to make the TXIDs unique
            vicky.pubkey,
            // OP_CHECKSIGVERIFY,
            // paul.pubkey,
            OP_CHECKSIG
        ]
    }

    unlock(vicky, paul, merkleIndex) {
        return [
            // paul.sign(this), // TODO
            vicky.sign(this),
        ]
    }
}

export class MerkleChallengeC extends Transaction {
    static ACTOR = VICKY
    static taproot(params) {
        return [
            [MerkleChallengeCLeaf, params.vicky, params.paul, this.INDEX]
        ]
    }
}


export class MerkleChallengeCTimeoutLeaf extends TimeoutLeaf {

    lock(vicky, paul) {
        return [
            TIMEOUT,
            OP_CHECKSEQUENCEVERIFY,
            OP_DROP,
            paul.pubkey,
            OP_CHECKSIG,
        ]
    }

    unlock(vicky, paul) {
        return [
            paul.sign(this),
        ]
    }
}



export class MerkleChallengeCTimeout extends EndTransaction {
    static ACTOR = PAUL
    static taproot(state) {
        return [
            [MerkleChallengeCTimeoutLeaf, state.vicky, state.paul]
        ]
    }
}



export class MerkleResponseCLeaf extends Leaf {

    lock(vicky, paul, merkleIndex) {
        return [
            paul.commit.merkleResponseCNextSibling(merkleIndex),
            paul.commit.merkleResponseCNext(merkleIndex),
            // vicky.pubkey,
            // OP_CHECKSIGVERIFY,
            paul.pubkey,
            OP_CHECKSIG,
        ]
    }

    unlock(vicky, paul, merkleIndex) {
        return [
            paul.sign(this),
            // vicky.sign(this),
            paul.unlock.merkleResponseCNext(merkleIndex),
            paul.unlock.merkleResponseCNextSibling(merkleIndex),
        ]
    }
}

export class MerkleResponseC extends Transaction {
    static ACTOR = PAUL
    static taproot(params) {
        return [
            [MerkleResponseCLeaf, params.vicky, params.paul, this.INDEX]
        ]
    }
}




export class MerkleResponseCTimeoutLeaf extends TimeoutLeaf {

    lock(vicky, paul) {
        return [
            TIMEOUT,
            OP_CHECKSEQUENCEVERIFY,
            OP_DROP,
            vicky.pubkey,
            OP_CHECKSIG,
        ]
    }

    unlock(vicky, paul) {
        return [
            vicky.sign(this),
        ]
    }
}


export class MerkleResponseCTimeout extends EndTransaction {
    static ACTOR = VICKY
    static taproot(state) {
        return [
            [MerkleResponseCTimeoutLeaf, state.vicky, state.paul]
        ]
    }
}





export class MerkleHashCLeftLeaf extends Leaf {

    lock(vicky, paul, merkleIndexC) {
        return [
            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            paul.push.addressCBitAt(PATH_LEN - 1 - merkleIndexC),
            OP_NOT,
            OP_VERIFY,

            // Read the child node
            paul.push.merkleResponseCNext(merkleIndexC),
            // Read the child's sibling
            u160_toaltstack,
            paul.push.merkleResponseCNextSibling(merkleIndexC),
            u160_fromaltstack,

            // Hash the child nodes
            blake3_160,
            u160_toaltstack,
            // Read the parent hash
            paul.push.merkleResponseCNext(merkleIndexC + 1),

            u160_fromaltstack,
            u160_swap_endian,
            u160_equalverify,
            OP_TRUE, // TODO: verify the covenant here
        ]
    }

    unlock(vicky, paul, merkleIndexC) {
        return [
            paul.unlock.merkleResponseCNext(merkleIndex),
            paul.unlock.merkleResponseCNextSibling(merkleIndex),
            paul.unlock.merkleResponseCNext(merkleIndex + 1),
            paul.unlock.addressCBitAt(PATH_LEN - 1 - merkleIndexC),
        ]
    }
}



export class MerkleHashCRightLeaf extends Leaf {

    lock(vicky, paul, merkleIndexC) {
        return [
            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            paul.push.addressCBitAt(PATH_LEN - 1 - merkleIndexC),
            OP_VERIFY,

            // Read the child's sibling
            paul.push.merkleResponseCNextSibling(merkleIndexC),
            // Read the child node
            u160_toaltstack,
            paul.push.merkleResponseCNext(merkleIndexC),
            u160_fromaltstack,

            // Hash the child nodes
            blake3_160,
            u160_toaltstack,
            // Read the parent hash
            paul.push.merkleResponseCNext(merkleIndexC + 1),

            u160_fromaltstack,
            u160_swap_endian,
            u160_equalverify,
            OP_TRUE, // TODO: verify the covenant here
        ]
    }

    unlock(vicky, paul, merkleIndexC) {
        return [
            paul.unlock.merkleResponseCNextSibling(merkleIndex),
            paul.unlock.merkleResponseCNext(merkleIndex),
            paul.unlock.merkleResponseCNext(merkleIndex + 1),
            paul.unlock.addressCBitAt(PATH_LEN - 1 - merkleIndexC),
        ]
    }
}



export class MerkleHashCRootLeftLeaf extends Leaf {

    lock(vicky, paul, traceRoundIndex) {
        return [
            // Verify we're executing the correct leaf

            vicky.push.traceIndex,
            OP_TOALTSTACK,
            vicky.push.nextTraceIndex(traceRoundIndex),
            OP_FROMALTSTACK,
            OP_EQUALVERIFY,

            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            paul.push.addressCBitAt(PATH_LEN - 1),
            OP_NOT,
            OP_VERIFY,

            // Read the child nodes
            paul.push.merkleResponseCNext(PATH_LEN - 1),
            // Read the child's sibling
            u160_toaltstack,
            paul.push.merkleResponseCNextSibling(PATH_LEN - 1),
            u160_fromaltstack,

            // Hash the child nodes
            blake3_160,
            u160_toaltstack,
            // Read the parent hash
            paul.push.traceResponse(traceRoundIndex),

            u160_fromaltstack,
            u160_swap_endian,
            u160_equalverify,

            OP_TRUE, // TODO: verify the covenant here
        ]
    }

    unlock(vicky, paul, traceRoundIndex) {
        return [
            paul.unlock.traceResponse(traceRoundIndex),
            paul.unlock.merkleResponseCNextSibling(PATH_LEN - 1),
            paul.unlock.merkleResponseCNext(PATH_LEN - 1),
            paul.unlock.addressCBitAt(PATH_LEN - 1),
            vicky.unlock.nextTraceIndex(traceRoundIndex),
            vicky.unlock.traceIndex,
        ]
    }
}



export class MerkleHashCRootRightLeaf extends Leaf {

    lock(vicky, paul, traceRoundIndex) {
        return [
            // Verify we're executing the correct leaf

            vicky.push.traceIndex,
            OP_TOALTSTACK,
            vicky.push.nextTraceIndex(traceRoundIndex),
            OP_FROMALTSTACK,
            OP_EQUALVERIFY,

            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            paul.push.addressCBitAt(PATH_LEN - 1),
            OP_VERIFY,

            // Read the child's sibling
            paul.push.merkleResponseCNextSibling(PATH_LEN - 1),
            // Read the child nodes
            paul.push.merkleResponseCNext(PATH_LEN - 1),
            u160_toaltstack,
            u160_fromaltstack,

            // Hash the child nodes
            blake3_160,
            u160_toaltstack,
            // Read the parent hash
            paul.push.traceResponse(traceRoundIndex),

            u160_fromaltstack,
            u160_swap_endian,
            u160_equalverify,

            OP_TRUE, // TODO: verify the covenant here
        ]
    }

    unlock(vicky, paul, traceRoundIndex) {
        return [
            paul.unlock.traceResponse(traceRoundIndex),
            paul.unlock.merkleResponseCNext(PATH_LEN - 1),
            paul.unlock.merkleResponseCNextSibling(PATH_LEN - 1),
            paul.unlock.addressCBitAt(PATH_LEN - 1),
            vicky.unlock.nextTraceIndex(traceRoundIndex),
            vicky.unlock.traceIndex, // TODO: Vicky can equivocate here
        ]
    }
}




export class MerkleCLeafHashLeftLeaf extends Leaf {

    lock(vicky, paul) {
        return [

            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            paul.push.addressCBitAt(0),
            OP_NOT,
            OP_VERIFY,

            // Read valueC
            paul.push.valueC,
            // Pad with 16 zero bytes
            u32_toaltstack,
            loop(16, _ => 0),
            u32_fromaltstack,
            
            // Read sibling
            u160_toaltstack,
            paul.push.merkleResponseCNextSibling(0),
            u160_fromaltstack,

            // Hash the child nodes
            blake3_160,
            u160_toaltstack,
            // Read the parent hash
            paul.push.merkleResponseCNext(LOG_PATH_LEN - 1),

            u160_fromaltstack,
            u160_swap_endian,
            u160_equalverify,
            OP_TRUE, // TODO: verify the covenant here
        ]
    }

    unlock(vicky, paul) {
        return [
            paul.unlock.merkleResponseCNext(1),
            paul.unlock.merkleResponseCNextSibling(0),
            paul.unlock.valueC,
            paul.unlock.addressCBitAt(0),
        ]
    }
}

export class MerkleCLeafHashRightLeaf extends Leaf {

    lock(vicky, paul) {
        return [
            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            paul.push.addressCBitAt(0),
            OP_VERIFY,

            // Read sibling
            paul.push.merkleResponseCNextSibling(0),
            
            // Read valueC
            u160_toaltstack,
            paul.push.valueC,
            // Pad with 16 zero bytes
            u32_toaltstack,
            loop(16, _ => 0),
            u32_fromaltstack,
            u160_fromaltstack,
            
            // Hash the child nodes
            blake3_160,
            u160_toaltstack,
            // Read the parent hash
            paul.push.merkleResponseCNext(LOG_PATH_LEN - 1),

            u160_fromaltstack,
            u160_swap_endian,
            u160_equalverify,
            OP_TRUE, // TODO: verify the covenant here
        ]
    }

    unlock(vicky, paul) {
        return [
            paul.unlock.merkleResponseCNext(1),
            paul.unlock.valueC,
            paul.unlock.merkleResponseCNextSibling(0),
            paul.unlock.addressCBitAt(0),
        ]
    }
}



export class MerkleHashC extends Transaction {
    static ACTOR = PAUL
    static taproot(params) {
        const { vicky, paul } = params;
        switch (this.INDEX) {
            case 0:
                return [
                    [MerkleCLeafHashLeftLeaf, vicky, paul],
                    [MerkleCLeafHashRightLeaf, vicky, paul],
                ];
            case (PATH_LEN - 1):
                return [
                    ...loop(LOG_TRACE_LEN, traceRoundIndex => [MerkleHashCRootLeftLeaf, vicky, paul, traceRoundIndex]),
                    ...loop(LOG_TRACE_LEN, traceRoundIndex => [MerkleHashCRootRightLeaf, vicky, paul, traceRoundIndex]),
                ];
            default:
                return [
                    [MerkleHashCLeftLeaf, vicky, paul, this.INDEX],
                    [MerkleHashCRightLeaf, vicky, paul, this.INDEX],
                ];
        }
    }
}



export class MerkleEquivocationC extends EndTransaction {
    static ACTOR = VICKY

    static taproot(params) {
        console.warn(`${this.name} not implemented`)
        return [
            [class extends Leaf {
                lock() {
                    return ['OP_4']
                }
                unlock() {
                    return []
                }
            }]
        ]
    }
}

File: bitvm/src/merkle/write.rs
use std::vec;

use scripts::leaf::Leaves;
use scripts::opcodes::pseudo::OP_CHECKSEQUENCEVERIFY;
use scripts::{opcodes::pushable, leaf::Leaf};
use bitcoin_script::bitcoin_script as script;
use scripts::opcodes::blake3::blake3_160;
use scripts::opcodes::{
    unroll,
    u160_std::{
        u160_fromaltstack,
        u160_toaltstack,
        u160_equalverify,
        u160_swap_endian,
    },
    u32_std::{
        u32_fromaltstack,
        u32_toaltstack,
    },
};
use crate::graph::BitVmLeaf;
use crate::model::BitVmModel;
use crate::constants::{LOG_PATH_LEN, PATH_LEN, TIMEOUT};

fn merkle_challenge_c_leaf<const MERKLE_INDEX: u8>() -> BitVmLeaf {
    BitVmLeaf {
        lock: |model| script! {
            { MERKLE_INDEX }
            OP_DROP // This is just a marker to make the TXIDs unique
            // { model.vicky.pubkey, }
            OP_CHECKSIGVERIFY
            // paul.pubkey
            OP_CHECKSIG
        },
        unlock: |model| script! {
            // paul.sign(this), // TODO
            // { model.vicky.sign(this), }
        }
    }
}

pub fn merkle_challenge_c<const MERKLE_INDEX: u8>() -> Vec<BitVmLeaf> {
    vec![
        merkle_challenge_c_leaf::<MERKLE_INDEX>()
    ]
}


fn merkle_challenge_c_timeout_leaf() -> BitVmLeaf {
    BitVmLeaf {
        lock: |model| script! {
            { TIMEOUT }
            OP_CHECKSEQUENCEVERIFY
            OP_DROP
            // { model.paul.pubkey }
            OP_CHECKSIG
        },
        unlock: |model| script! {
            // { model.paul.sign(this) }
        }
    }
}



// // export class MerkleChallengeCTimeout extends EndTransaction {
// //     static ACTOR = PAUL
// //     static taproot(state) -> Script {
// //         return [
// //             [MerkleChallengeCTimeoutLeaf, state.vicky, state.paul]
// //         ]
// //     }
// // }


fn merkle_response_c_leaf<const MERKLE_INDEX: u8>() -> BitVmLeaf {
    BitVmLeaf {
        lock: |model| script! {
            { model.paul.commit().merkle_response_c_next_sibling(MERKLE_INDEX) }
            { model.paul.commit().merkle_response_c_next(MERKLE_INDEX) }
            // vicky.pubkey
            OP_CHECKSIGVERIFY
            // { model.paul.pubkey }
            OP_CHECKSIG
        },
        unlock: |model| script! {
            // paul.sign(this)
            // vicky.sign(this)
            { model.paul.unlock().merkle_response_c_next(MERKLE_INDEX) }
            { model.paul.unlock().merkle_response_c_next_sibling(MERKLE_INDEX) }
        }
    }
}

// // export class MerkleResponseC extends Transaction {
// //     static ACTOR = PAUL
// //     static taproot(model) -> Script {
// //         return [
// //             [MerkleResponseCLeaf, model.vicky, model.paul, this.INDEX]
// //         ]
// //     }
// // }


fn merkle_response_c_timeout_leaf() -> BitVmLeaf {
    BitVmLeaf {
        lock: |model| script! {
            { TIMEOUT }
            OP_CHECKSEQUENCEVERIFY
            OP_DROP
            // { self.vicky.pubkey }
            OP_CHECKSIG
        },
        unlock: |model| script! {
            // { self.vicky.sign(this) }
        }
    }
}

// // export class MerkleResponseCTimeout extends EndTransaction {
// //     static ACTOR = VICKY
// //     static taproot(state) -> Script {
// //         return [
// //             [MerkleResponseCTimeoutLeaf, state.vicky, state.paul]
// //         ]
// //     }
// // }


fn merkle_hash_c_left_leaf<const MERKLE_INDEX: u8>() -> BitVmLeaf {
    BitVmLeaf {
        lock: |model| script! {
            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            { model.paul.push().address_c_bit_at(PATH_LEN as u8 - 1 - MERKLE_INDEX) }
            OP_NOT
            OP_VERIFY

            // Read the child node
            { model.paul.push().merkle_response_c_next(MERKLE_INDEX) }
            // Read the child's sibling
            u160_toaltstack
            { model.paul.push().merkle_response_c_next_sibling(MERKLE_INDEX) }
            u160_fromaltstack

            // Hash the child nodes
            blake3_160
            u160_toaltstack
            // Read the parent hash
            { model.paul.push().merkle_response_c_next(MERKLE_INDEX + 1) }

            u160_fromaltstack
            u160_swap_endian
            u160_equalverify
            OP_TRUE // TODO: verify the covenant here
        },
        unlock: |model| script! {
            { model.paul.unlock().merkle_response_c_next(MERKLE_INDEX) }
            { model.paul.unlock().merkle_response_c_next_sibling(MERKLE_INDEX) }
            { model.paul.unlock().merkle_response_c_next(MERKLE_INDEX + 1) }
            { model.paul.unlock().address_c_bit_at(PATH_LEN as u8 - 1 - MERKLE_INDEX) }
        }
    }
}

fn merkle_hash_c_right_leaf<const MERKLE_INDEX_C: u8>() -> BitVmLeaf {
    BitVmLeaf {
        lock: |model| script! {
                // Read the bit from address to figure out if we have to swap the two nodes before hashing
            { model.paul.push().address_c_bit_at(PATH_LEN as u8 - 1 - MERKLE_INDEX_C) }
            OP_VERIFY

            // Read the child's sibling
            { model.paul.push().merkle_response_c_next_sibling(MERKLE_INDEX_C) }
            // Read the child node
            u160_toaltstack
            { model.paul.push().merkle_response_c_next(MERKLE_INDEX_C) }
            u160_fromaltstack

            // Hash the child nodes
            blake3_160
            u160_toaltstack
            // Read the parent hash
            { model.paul.push().merkle_response_c_next(MERKLE_INDEX_C + 1) }

            u160_fromaltstack
            u160_swap_endian
            u160_equalverify
            OP_TRUE // TODO: verify the covenant here
        },
        
        unlock: |model| script! {
            { model.paul.unlock().merkle_response_c_next_sibling(MERKLE_INDEX_C) }
            { model.paul.unlock().merkle_response_c_next(MERKLE_INDEX_C) }
            { model.paul.unlock().merkle_response_c_next(MERKLE_INDEX_C + 1) }
            { model.paul.unlock().address_c_bit_at(PATH_LEN as u8 - 1 - MERKLE_INDEX_C) }
        }
    }
}


fn merkle_hash_c_root_left_leaf<const TRACE_ROUND_INDEX: u8>() -> BitVmLeaf {
    BitVmLeaf {
        lock: |model| script! {
            // Verify we're executing the correct leaf

            { model.vicky.push().trace_index() }
            OP_TOALTSTACK
            { model.vicky.push().next_trace_index(TRACE_ROUND_INDEX) }
            OP_FROMALTSTACK
            OP_EQUALVERIFY

            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            { model.paul.push().address_c_bit_at(PATH_LEN as u8 - 1) }
            OP_NOT
            OP_VERIFY

            // Read the child nodes
            { model.paul.push().merkle_response_c_next(PATH_LEN as u8 - 1) }
            // Read the child's sibling
            u160_toaltstack
            { model.paul.push().merkle_response_c_next_sibling(PATH_LEN as u8 - 1) }
            u160_fromaltstack

            // Hash the child nodes
            blake3_160
            u160_toaltstack
            // Read the parent hash
            { model.paul.push().trace_response(TRACE_ROUND_INDEX) }

            u160_fromaltstack
            u160_swap_endian
            u160_equalverify

            OP_TRUE // TODO: verify the covenant here
        },

        unlock: |model| script! {
            { model.paul.unlock().trace_response(TRACE_ROUND_INDEX) }
            { model.paul.unlock().merkle_response_c_next_sibling(PATH_LEN as u8 - 1) }
            { model.paul.unlock().merkle_response_c_next(PATH_LEN as u8 - 1) }
            { model.paul.unlock().address_c_bit_at(PATH_LEN as u8 - 1) }
            { model.vicky.unlock().next_trace_index(TRACE_ROUND_INDEX) }
            { model.vicky.unlock().trace_index() }
        }
    }
}


fn merkle_hash_c_root_right_leaf<const TRACE_ROUND_INDEX: u8>() -> BitVmLeaf {
    BitVmLeaf {
        lock: |model| script! {
            // Verify we're executing the correct leaf

            { model.vicky.push().trace_index() }
            OP_TOALTSTACK
            { model.vicky.push().next_trace_index(TRACE_ROUND_INDEX) }
            OP_FROMALTSTACK
            OP_EQUALVERIFY

            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            { model.paul.push().address_c_bit_at(PATH_LEN as u8 - 1) }
            OP_VERIFY

            // Read the child's sibling
            { model.paul.push().merkle_response_c_next_sibling(PATH_LEN as u8 - 1) }
            // Read the child nodes
            { model.paul.push().merkle_response_c_next(PATH_LEN as u8 - 1) }
            u160_toaltstack
            u160_fromaltstack

            // Hash the child nodes
            blake3_160
            u160_toaltstack
            // Read the parent hash
            { model.paul.push().trace_response(TRACE_ROUND_INDEX) }

            u160_fromaltstack
            u160_swap_endian
            u160_equalverify

            OP_TRUE // TODO: verify the covenant here
        },

        unlock: |model| script! {
            { model.paul.unlock().trace_response(TRACE_ROUND_INDEX) }
            { model.paul.unlock().merkle_response_c_next(PATH_LEN as u8 - 1) }
            { model.paul.unlock().merkle_response_c_next_sibling(PATH_LEN as u8 - 1) }
            { model.paul.unlock().address_c_bit_at(PATH_LEN as u8 - 1) }
            { model.vicky.unlock().next_trace_index(TRACE_ROUND_INDEX) }
            { model.vicky.unlock().trace_index() } // TODO: Vicky can equivocate here
        }
    }
}

const MERKLE_CLEAF_HASH_LEFT_LEAF: BitVmLeaf = BitVmLeaf {
    lock: |model| script! {
    
        // Read the bit from address to figure out if we have to swap the two nodes before hashing
        { model.paul.push().address_c_bit_at(0) }
        OP_NOT
        OP_VERIFY
    
        // Read value_c
        { model.paul.push().value_c() }
        // Pad with 16 zero bytes
        u32_toaltstack
        { unroll(16, |_| 0) }
        u32_fromaltstack
        
        // Read sibling
        u160_toaltstack
        { model.paul.push().merkle_response_c_next_sibling(0) }
        u160_fromaltstack
    
        // Hash the child nodes
        blake3_160
        u160_toaltstack
        // Read the parent hash
        { model.paul.push().merkle_response_c_next(LOG_PATH_LEN as u8 - 1) }
    
        u160_fromaltstack
        u160_swap_endian
        u160_equalverify
        OP_TRUE // TODO: verify the covenant here
    },
    unlock: |model| script! {
        { model.paul.unlock().merkle_response_c_next(1) }
        { model.paul.unlock().merkle_response_c_next_sibling(0) }
        { model.paul.unlock().value_c() }
        { model.paul.unlock().address_c_bit_at(0) }
    }
};

const MERKLE_C_LEAF_HASH_RIGHT_LEAF: BitVmLeaf = BitVmLeaf {
    lock: |model| {
        script! {
            // Read the bit from address to figure out if we have to swap the two nodes before hashing
            { model.paul.push().address_c_bit_at(0) }
            OP_VERIFY

            // Read sibling
            { model.paul.push().merkle_response_c_next_sibling(0) }
            
            // Read value_c
            u160_toaltstack
            { model.paul.push().value_c() }
            // Pad with 16 zero bytes
            u32_toaltstack
            { unroll(16, |_| 0) }
            u32_fromaltstack
            u160_fromaltstack
            
            // Hash the child nodes
            blake3_160
            u160_toaltstack
            // Read the parent hash
            { model.paul.push().merkle_response_c_next(LOG_PATH_LEN as u8 - 1) }

            u160_fromaltstack
            u160_swap_endian
            u160_equalverify
            OP_TRUE // TODO: verify the covenant here
        }
    },

    unlock: |model| script! {
        { model.paul.unlock().merkle_response_c_next(1) }
        { model.paul.unlock().value_c() }
        { model.paul.unlock().merkle_response_c_next_sibling(0) }
        { model.paul.unlock().address_c_bit_at(0) }
    }
};



// // export class MerkleHashC extends Transaction {
// //     static ACTOR = PAUL
// //     static taproot(model) -> Script {
// //         const { vicky, paul } = model;
// //         switch (this.INDEX) -> Script {
// //             case 0:
// //                 return [
// //                     [MerkleCLeafHashLeftLeaf, vicky, paul],
// //                     [MerkleCLeafHashRightLeaf, vicky, paul],
// //                 ];
// //             case (PATH_LEN as u8 - 1):
// //                 return [
// //                     ...loop(LOG_TRACE_LEN, traceRoundIndex => [MerkleHashCRootLeftLeaf, vicky, paul, traceRoundIndex]),
// //                     ...loop(LOG_TRACE_LEN, traceRoundIndex => [MerkleHashCRootRightLeaf, vicky, paul, traceRoundIndex]),
// //                 ];
// //             default:
// //                 return [
// //                     [MerkleHashCLeftLeaf, vicky, paul, this.INDEX],
// //                     [MerkleHashCRightLeaf, vicky, paul, this.INDEX],
// //                 ];
// //         }
// //     }
// // }



// // export class MerkleEquivocationC extends EndTransaction {
// //     static ACTOR = VICKY

// //     static taproot(model) -> Script {
// //         console.warn(`${this.name} not implemented`)
// //         return [
// //             [class extends Leaf<'a> {
// //                 lock() -> Script {
// //                     return ['OP_4']
// //                 }
// //                 unlock() -> Script {
// //                     return []
// //                 }
// //             }]
// //         ]
// //     }
// // }


